File: Sources/iOS/Presenter Support/View/UIViewController+PresenterSupport.m
//
//  Copyright © 2016 Cleverlance. All rights reserved.
//

#import "UIViewController+PresenterSupport.h"
#import <Pyramid/Pyramid-Swift.h>
#import <objc/runtime.h>

@implementation UIViewController (PresenterSupport)

#pragma mark - Swizzling

+ (void)load {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        [self swizzleOriginalSelector:@selector(viewWillAppear:)
                     swizzledSelector:@selector(customViewWillAppear:)
                              onClass:[self class]];

        [self swizzleOriginalSelector:@selector(viewDidDisappear:)
                     swizzledSelector:@selector(customViewDidDisappear:)
                              onClass:[self class]];
    });
}


+ (void)swizzleOriginalSelector:(SEL)originalSelector
               swizzledSelector:(SEL)swizzledSelector
                        onClass:(Class)class
{
    Method originalMethod = class_getInstanceMethod(class, originalSelector);
    Method swizzledMethod = class_getInstanceMethod(class, swizzledSelector);

    BOOL didAddMethod = class_addMethod(class,
                                        originalSelector,
                                        method_getImplementation(swizzledMethod),
                                        method_getTypeEncoding(swizzledMethod));

    if (didAddMethod) {
        class_replaceMethod(class,
                            swizzledSelector,
                            method_getImplementation(originalMethod),
                            method_getTypeEncoding(originalMethod));
    } else {
        method_exchangeImplementations(originalMethod, swizzledMethod);
    }
}
@end

File: Sources/iOS/Presenter Support/View/UIViewController+PresenterSupport.swift
//
//  Copyright © 2016 Cleverlance. All rights reserved.
//

import UIKit

extension UIViewController {

    @objc public func customViewWillAppear(_ animated: Bool) {
        self.customViewWillAppear(animated)

        if let holder = self as? MultiplePresenterHolder {
            holder.presenters.forEach { $0.isActive = true }
        }
    }

    @objc public func customViewDidDisappear(_ animated: Bool) {
        self.customViewDidDisappear(animated)

        if let holder = self as? MultiplePresenterHolder {
            holder.presenters.forEach { $0.isActive = false }
        }
    }
}
